src/Message/ReportingAbstractRequest.php

```
@@ -0,0 +1,69 @@
+<?php
+
+
+namespace Omnipay\BlueSnap\Message;
+
+
+use Omnipay\BlueSnap\Constants;
+use Omnipay\Common\Exception\InvalidRequestException;
+
+/**
+ * Abstract class for the BlueSnap Reporting API
+ *
+ * @link https://developers.bluesnap.com/v8976-Tools/docs/bluesnap-reporting-tools
+ *
+ * @package Omnipay\BlueSnap\Message
+ */
+abstract class ReportingAbstractRequest extends AbstractRequest
+{
+    /**
+     * @return string
+     */
+    abstract public function getReportName();
+
+    /**
+     * Returns the HTTP method to be used for this request
+     *
+     * @return string
+     */
+    public function getHttpMethod()
+    {
+        return Constants::HTTP_METHOD_GET;
+    }
+
+    /**
+     * @return null
+     * @throws InvalidRequestException
+     */
+    public function getData()
+    {
+        $this->validate('startTime');
+        $this->validate('endTime');
+
+        if ($this->getStartTime() > $this->getEndTime()) {
+            throw new InvalidRequestException('startTime cannot be greater than endTime');
+        }
+
+        return null;
+    }
+
+    /**
+     * Return the API endpoint.
+     *
+     * @return string
+     */
+    public function getEndpoint()
+    {
+        $startTime = $this->getStartTime();
+        $endTime = $this->getEndTime();
+        $endpoint = parent::getEndpoint() . '/report/' . $this->getReportName();
+
+        // this should always be true
```

in that case should we throw an error when it's not? what will happen if start and end time aren't set - will the request time out trying to retrieve too many records?

---

I'm not sure. This is existing code but I can throw an exception if values aren't set